trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/expressiontranslation/BitabsTranslation.java

```diff
@@ -30,384 +30,360 @@
-import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.HeapLValue;
-import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.LRValue;
-import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.Result;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.RValue;
 import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.util.SFO;
-import de.uni_freiburg.informatik.ultimate.cdt.translation.interfaces.handler.ITypeHandler;
+import de.uni_freiburg.informatik.ultimate.core.lib.models.annotation.Overapprox;
 import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+import de.uni_freiburg.informatik.ultimate.util.datastructures.relation.Pair;
 
 /**
+ * This class is used to overapproximate bitwise operations for the integer translation.
  *
  * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
  * @author Cyrus Liu (yliu195@stevens.edu)
  *
  */
 public class BitabsTranslation {
-	private final FunctionDeclarations mFunctionDeclarations;
 	private final TypeSizes mTypeSizes;
-	private final ITypeHandler mTypeHandler;
 
-	public BitabsTranslation(final TypeSizes typeSizes, final ITypeHandler typeHandler,
-			final FunctionDeclarations functionDeclarations) {
+	public BitabsTranslation(final TypeSizes typeSizes) {
 		mTypeSizes = typeSizes;
-		mTypeHandler = typeHandler;
-		mFunctionDeclarations = functionDeclarations;
 	}
 
-	public Expression abstractAnd(final ILocation loc, final Expression left, final CPrimitive typeLeft,
-			final Expression right, final CPrimitive typeRight) {
+	/**
+	 * Overapproximates the bitwise {@code and}. Uses the following rules to increase the precision:
+	 * <li>0 & a = a & 0 = 0
+	 * <li>a & a = a
+	 * <li>If a >= 0 or b < 0, then a & b <= a
+	 * <li>If a < 0 or b >= 0, then a & b <= b
+	 * <li>If a >= b or b >= 0, then a & b >= 0
+	 * <li>If a < 0 or b < 0, then a & b > a + b
+	 */
+	public ExpressionResult abstractAnd(final ILocation loc, final Expression left, final Expression right,
+			final CPrimitive type, final AuxVarInfoBuilder auxVarInfoBuilder) {
 		// 0 & a = a & 0 = 0
-		if (left instanceof IntegerLiteral && ((IntegerLiteral) left).getValue().equals("0")) {
-			return left;
+		if (isZero(left)) {
+			return new ExpressionResult(new RValue(left, type));
 		}
-		if (right instanceof IntegerLiteral && ((IntegerLiteral) right).getValue().equals("0")) {
-			return right;
+		if (isZero(right)) {
+			return new ExpressionResult(new RValue(right, type));
 		}
-		// a & a = a
-		if (left instanceof IntegerLiteral && right instanceof IntegerLiteral
-				&& ((IntegerLiteral) left).getValue().equals(((IntegerLiteral) right).getValue())) {
-			return left;
+		if (left instanceof IntegerLiteral && right instanceof IntegerLiteral) {
+			return handleConstants((IntegerLiteral) left, (IntegerLiteral) right, BigInteger::and, loc, type);
```

Comment:
I wonder if it would be worth here to use `TypeSizes::extractIntegerValue` here, for expressions that can be evaluated though they are not directly literals? Perhaps in combination with adding support for evaluating bitwise operations in `ValueExtraction::extractIntegerValueFromBitvectorFunctionApplication` -- that way, you could handle also expressions like `0x00F & (C | 0xF00)` (where `C` is a constant) precisely.

Reply:
I just talked to @Heizmann about this. We want to get rid of `TypeSizes::extractIntegerValue` and directly use the literals everywhere instead of the constants. With these changes this code seems fine.